torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -121,6 +122,8 @@ Reducer::Reducer(
                 torch::make_unique<torch::autograd::utils::LambdaPostHook>(
                     [=](const torch::autograd::variable_list& outputs,
                         const torch::autograd::variable_list& /* unused */) {
+                      this->rpc_context_.set(
```

Comment:
what if RPC is not initialized?

Reply:
> what if RPC is not initialized?

the weak ptr from the thread local variable will be nullptr.

Reply:
nit: I'm wondering whether it needs to set the context inside every grad variable's hook and every iteration? maybe just set once, because it is thread local variable?

Reply:
> nit: I'm wondering whether it needs to set the context inside every grad variable's hook and every iteration? maybe just set once, because it is thread local variable?

@zhaojuanmao Is there a way to tell the end of an iteration? Because the context may change iteration by iteration. E.g.

```
m = DDP(model)
with dist_autograd.context() as context_id:
    # do backward() and allreduce()

# The reducer instance has the previous context id
with dist_autograd.context() as context_id:
    # The reducer instance has to know the new context id.
    # do backward() and allreduce()

```

Reply:
Now I clear the context in `Reducer::finalize_backward()`, which seems to indicate the end of a backward run.